frame/undercollateralized-loans/src/helpers.rs

```diff
@@ -0,0 +1,529 @@
+use crate::{
+	types::{
+		LoanConfigOf, LoanInputOf, MarketConfigOf, MarketInfoOf, MarketInputOf, RepaymentResult,
+		Timestamp,
+	},
+	validation::{AssetIsSupportedByOracle, CurrencyPairIsNotSame, LoanInputIsValid},
+	Config, DebtTokenForMarketStorage, Error, MarketsStorage, Pallet,
+};
+use composable_support::validation::Validated;
+use composable_traits::{
+	currency::CurrencyFactory,
+	defi::DeFiComposableConfig,
+	oracle::Oracle,
+	undercollateralized_loans::{LoanConfig, MarketConfig, MarketInfo, UndercollateralizedLoans},
+	vault::{Deposit, FundsAvailability, StrategicVault, Vault, VaultConfig},
+};
+use frame_support::{
+	ensure,
+	storage::with_transaction,
+	traits::{
+		fungible::{Inspect as NativeInspect, Transfer as NativeTransfer},
+		fungibles::{Inspect, Mutate, Transfer},
+		Get, UnixTime,
+	},
+};
+use sp_runtime::{
+	traits::{One, Saturating, Zero},
+	DispatchError, Perquintill, TransactionOutcome,
+};
+
+use chrono::{NaiveDate, NaiveDateTime, NaiveTime};
+use sp_std::vec::Vec;
+
+// #generalization
+impl<T: Config> Pallet<T> {
+	pub(crate) fn do_create_market(
+		manager: T::AccountId,
+		input: Validated<
+			MarketInputOf<T>,
+			(CurrencyPairIsNotSame, AssetIsSupportedByOracle<T::Oracle>),
+		>,
+		keep_alive: bool,
+	) -> Result<MarketInfoOf<T>, DispatchError> {
+		let config_input = input.value();
+		crate::MarketsCounterStorage::<T>::try_mutate(|counter| {
+			*counter += T::Counter::one();
+			ensure!(
+				*counter <= T::MaxMarketsCounterValue::get(),
+				Error::<T>::ExceedMaxMarketsCounterValue
+			);
+			let market_account_id = Self::market_account_id(*counter);
+			let borrow_asset_vault = T::Vault::create(
+				Deposit::Existential,
+				VaultConfig {
+					asset_id: config_input.borrow_asset(),
+					reserved: config_input.reserved_factor(),
+					manager: manager.clone(),
+					strategies: [(
+						market_account_id.clone(),
+						Perquintill::one().saturating_sub(config_input.reserved_factor()),
+					)]
+					.into_iter()
+					.collect(),
+				},
+			)?;
+
+			let initial_pool_size = Self::calculate_initial_pool_size(config_input.borrow_asset())?;
+
+			ensure!(
+				initial_pool_size > T::Balance::zero(),
+				Error::<T>::PriceOfInitialBorrowVaultShouldBeGreaterThanZero
+			);
+
+			T::MultiCurrency::transfer(
```

Comment:
Is this the only time it's possible to add the borrowable asset into the market? We would want a window in which we could publicly announce the opening of a new market (with a predetermined interest/payment schedule) in which users could send the borrowable assets into the market.

Reply:
@mikolaichuk lenders interact with the market through the vault. So, they always can use https://github.com/ComposableFi/composable/blob/mck/undercollateralized-loans-schedule-sb/frame/vault/src/lib.rs#L546. This part of code is used to deposit initial amount of money, which should be done once by market's manager.